doc/man/man1/scalac.1
.\" ##########################################################################
.\" #                      __                                                #
.\" #      ________ ___   / /  ___     Scala On-line Manual Pages            #
.\" #     / __/ __// _ | / /  / _ |    (c) 2002-2005, LAMP/EPFL              #
.\" #   __\ \/ /__/ __ |/ /__/ __ |                                          #
.\" #  /____/\___/_/ |_/____/_/ | |    http://scala.epfl.ch/                 #
.\" #                           |/                                           #
.\" ##########################################################################
.\"
.\" Process this file with nroff -man scalac.1
.\"
.TH scalac 1  "April 29, 2005" "version 0.1" "USER COMMANDS"
.\"
.\" ################################# NAME ###################################
.\"
.SH NAME
scalac \- Compiler for the Scala language
.\"
.\" ############################### SYNOPSIS #################################
.\"
.SH SYNOPSIS
\fBscalac\fR [ <\fIoptions\fR> ] <\fIsource files\fR>
.\"
.\" ############################### PARAMETERS ###############################
.\"
.SH PARAMETERS
.TP
<\fIoptions\fR>
Command line options. See \fBOPTIONS\fR below.
.TP
<\fIsource files\fR>
One or more source files to be compiled (such as MyClass.scala).
.\"
.\" ############################## DESCRIPTION ###############################
.\"
.SH DESCRIPTION
The \fBscalac\fR tool reads class and object definitions, written in the
Scala programming language, and compiles them into bytecode class files.
.PP
By default, the compiler puts each class file in the same directory as its
source file. You can specify a separate destination directory with -d (see
\fBOPTIONS\fR, below).
.\"
.\" ############################### OPTIONS ##################################
.\"
.SH OPTIONS
The compiler has a set of standard options that are supported on the current
development environment and will be supported in future releases.  An
additional set of non-standard options are specific to the current virtual
machine implementation and are subject to change in the future.  Non-standard
options begin with \fB-X\fR.
.SS Standard Options
.TP
\fB\-g\fR
Generate debugging info
.TP
\fB\-nowarn\fR
Generate no warnings
.TP
\fB\-verbose\fR
Output messages about what the compiler is doing
.TP
\fB\-classpath\fR <\fIpath\fR>
Specify where to find user class files (on Unix-based systems
a colon-separated list of paths, on Windows-based systems, a
semicolon-separate list of paths). This does not override the
built-in ("boot") search path.
.TP
\fB\-sourcepath\fR <\fIpath\fR>
Specify where to find input source files.
.TP
\fB\-bootclasspath\fR <\fIpath\fR>
Override location of bootstrap class files (where to find the
standard built-in classes, such as "scala.List").
.TP
\fB\-extdirs\fR <\fIdirs\fR>
Override location of installed extensions.
.TP
\fB\-d\fR <\fIdirectory\fR>
Specify where to place generated class files.
.TP
\fB\-encoding\fR <\fIencoding\fR>
Specify character encoding used by source files.
.TP
\fB\-separate:\fR<\fIseparate\fR>
Read symbol files for separate compilation: (\fIyes\fR, \fIno\fR).
.TP
\fB\-target:\fR<\fItarget\fR>
Specify which backend to use (\fIjvm\fR).
.TP
\fB\-version\fR
Print product version and exit.
.TP
\fB\-? -help\fR
Print a synopsis of standard options.
.SS Non-Standard Options
.TP
\fB\-Xshortname\fR
Display short file names in error reports.
.TP
\fB\-Xmarkup\fR
No effect, only for compatibility.
.TP
\fB\-Xnewmatch\fR
New pattern matching.
.TP
\fB\-XpreserveWS\fR
Don't trim whitespace in XML literals.
.TP
\fB\-XrootClass\fR <\fIclass\fR>
Specify the root class of the global application.
.TP
\fB\-XappFile\fR <\fIfile\fR>
Specify the filename where to dump the whole-application inferred from the rootclass.
.TP
\fB\-XdotFile\fR <\fIfile\fR>
Specify the filename where to print the AST in graphviz dot format.
.TP
\fB\-Xrta\fR
Perform Rapid Type analysis for method call resolution.
.TP
\fB\-Xinline\fR
Inline method calls that can be resolved statically.
.TP
\fB\-Xrtt\fR
Enable run-time types.
.SS Debug Options
.TP
.B \-debug
Output debugging messages.
.TP
\fB\-explaintypes\fR
Explain type errors in more detail.
.TP
\fB\-uniqid\fR
Print identifiers with unique names.
.TP
\fB\-types\fR
Print tree types.
.TP
\fB\-prompt\fR
Display a prompt after each error.
.TP
\fB\-noimports\fR
Compile without any implicit imports.
.TP
\fB\-nopredefs\fR
Compile without any implicit predefined values.
.TP
\fB\-skip:\fR<\fIphases\fR>
Skip <\fIphases\fR> (see below).
.TP
\fB\-check:\fR<\fIphases\fR>
Check the tree after <\fIphases\fR> (see below).
.TP
\fB\-print:\fR<phases>
Print out program after <\fIphases\fR> (see below).
.TP
\fB\-printer:\fR<\fIprinter\fR>
Printer to use.
.TP
\fB\-print-file\fR <\fIfile\fR>
Specify file in which to print trees.
.TP
\fB\-graph:\fR<\fIphases\fR>
Graph the program after <\fIphases\fR> (see below).
.TP
\fB\-stop:\fR<\fIphases\fR>
Stop after first phase in <\fIphases\fR> (see below).
.TP
\fB\-log:\fR<\fIphases\fR>
Log operations in <\fIphases\fR> (see below).
.SS Compilation Phases
.TP
\fIinitial\fR
initializing compiler
.TP
\fIparse\fR
parse source files
.TP
\fInamer\fR
create symbols
.TP
\fIanalyze\fR
name and type analysis
.TP
\fIrefcheck\fR
reference checking
.TP
\fIuncurry\fR
uncurry function types and applications
.TP
\fItransmatch\fR
translate match expressions
.TP
\fIlambdalift\fR
lambda lifter
.TP
\fItypesasvalues\fR
represent types as values
.TP
\fIaddaccessors\fR
add accessors for constructor arguments
.TP
\fIexplicitouterclasses\fR
make links from inner classes to enclosing one explicit
.TP
\fIaddconstructors\fR
add explicit constructor for each class
.TP
\fItailcall\fR
add tail-calls
.TP
\fIwholeprog\fR
perform whole program analysis
.TP
\fIaddinterfaces\fR
add one interface per class
.TP
\fIexpandmixins\fR
expand mixins by code copying
.TP
\fIboxing\fR
makes boxing explicit
.TP
\fIerasure\fR
type eraser
.TP
\fIicode\fR
generate icode
.TP
\fIcodegen\fR
enable code generation
.TP
\fIterminal\fR
compilation terminated
.TP
\fIall\fR
matches all phases

.\"
.\" ############################### EXAMPLES #################################
.\"
.SH EXAMPLES
.TP
Compile a Scala program
.BR
\fBscalac\fR HelloWorld
.TP
Compile a Scala program to the destination directory \fBclasses\fR
.BR
\fBscalac\fR \-d classes HelloWorld.scala
.TP
Compile all Scala files found in the source directory \fBsrc\fR to the
destination directory \fBclasses\fR
.BR
\fBscalac\fR \-d classes src/*.scala 
.\"
.\" ############################# EXIT STATUS ################################
.\"
.SH "EXIT STATUS"
\fBscalac\fR returns a zero exist status if it succeeds to compile the
specified input files. Non zero is returned in case of failure.
.\"
.\" ############################## AUTHOR(S) #################################
.\"
.SH AUTHOR
Written by Stephane Micheloud.
.\"
.\" ################################ BUGS ####################################
.\"
.SH "REPORTING BUGS"
Report bugs to <scala@lists.epfl.ch>.
.\"
.\" ############################# COPYRIGHT ##################################
.\"
.SH COPYRIGHT
Copyright \(co 2002-2005 LAMP/EPFL
.PP
This is free software; see the distribution for copying conditions.  There is
NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\" ############################## SEE ALSO ##################################
.\"
.SH "SEE ALSO"
.BR scala(1),
.BR scaladoc(1),
.BR scalaint(1),
.BR scalarun(1)
.\"
.\" ##########################################################################
.\" $Id$


src/compiler/scala/tools/ant/Scalac.scala
/*                     __                                               *\
**     ________ ___   / /  ___     Scala Ant Tasks                      **
**    / __/ __// _ | / /  / _ |    (c) 2005-2007, LAMP/EPFL             **
**  __\ \/ /__/ __ |/ /__/ __ |    http://scala-lang.org/               **
** /____/\___/_/ |_/____/_/ | |                                         **
**                          |/                                          **
\*                                                                      */

// $Id$

package scala.tools.ant

import java.io.File

import org.apache.tools.ant.{BuildException, Project}
import org.apache.tools.ant.taskdefs.MatchingTask
import org.apache.tools.ant.types.{Path, Reference}
import org.apache.tools.ant.util.{FileUtils, GlobPatternMapper,
                                  SourceFileScanner}

import scala.tools.nsc.{Global, Settings}
import scala.tools.nsc.reporters.{Reporter, ConsoleReporter}

/** <p>
 *    An Ant task to compile with the new Scala compiler (NSC).
 *  </p>
 *  <p>
 *    This task can take the following parameters as attributes:
 *  </p>
 *  <ul style="font-family:Courier;">
 *    <li>srcdir (mandatory),</li>
 *    <li>srcref,</li>
 *    <li>destdir,</li>
 *    <li>classpath,</li>
 *    <li>classpathref,</li>
 *    <li>sourcepath,</li>
 *    <li>sourcepathref,</li>
 *    <li>bootclasspath,</li>
 *    <li>bootclasspathref,</li>
 *    <li>extdirs,</li>
 *    <li>extdirsref,</li>
 *    <li>encoding,</li>
 *    <li>target,</li>
 *    <li>force,</li>
 *    <li>logging,</li>
 *    <li>logphase,</li>
 *    <li>usepredefs,</li>
 *    <li>debuginfo,</li>
 *    <li>addparams,</li>
 *    <li>scalacdebugging,</li>
 *    <li>deprecation,</li>
 *    <li>unchecked.</li>
 *  </ul>
 *  <p>
 *    It also takes the following parameters as nested elements:
 *  </p>
 *  <ul>
 *    <li>src (for srcdir),</li>
 *    <li>classpath,</li>
 *    <li>sourcepath,</li>
 *    <li>bootclasspath,</li>
 *    <li>extdirs.</li>
 * </ul>
 * @author Gilles Dubochet, Stephane Micheloud */
class Scalac extends MatchingTask {

  /** The unique Ant file utilities instance to use in this task. */
  private val fileUtils = FileUtils.newFileUtils()

/*============================================================================*\
**                             Ant user-properties                            **
\*============================================================================*/

  abstract class PermissibleValue {
    val values: List[String]
    def isPermissible(value: String): Boolean =
      (value == "") || values.exists(.startsWith(value))
  }

  /** Defines valid values for the logging property. */
  object LoggingLevel extends PermissibleValue {
    val values = List("none", "verbose", "debug")
  }

  /** Defines valid values for properties that refer to compiler phases. */
  object CompilerPhase extends PermissibleValue {
    val values = List("namer", "typer", "pickler", "uncurry", "tailcalls",
                      "transmatch", "explicitouter", "erasure", "lambdalift",
                      "flatten", "constructors", "mixin", "icode", "jvm",
                      "terminal")
  }

  /** Defines valid values for the <code>target</code> property. */
  object Target extends PermissibleValue {
    val values = List("jvm-1.5", "jvm-1.4", "msil", "cldc")
  }

  /** Defines valid values for the <code>deprecation</code> and
   *  <code>unchecked</code> properties. */
  object Flag extends PermissibleValue {
    val values = List("yes", "no", "on", "off")
  }

  /** The directories that contain source files to compile. */
  private var origin: Option[Path] = None
  /** The directory to put the compiled files in. */
  private var destination: Option[File] = None

  /** The class path to use for this compilation. */
  private var classpath: Option[Path] = None
  /** The source path to use for this compilation. */
  private var sourcepath: Option[Path] = None
  /** The boot class path to use for this compilation. */
  private var bootclasspath: Option[Path] = None
  /** The external extensions path to use for this compilation. */
  private var extdirs: Option[Path] = None

  /** The character encoding of the files to compile. */
  private var encoding: Option[String] = None

  // the targetted backend
  private var backend: Option[String] = None

  /** Whether to force compilation of all files or not. */
  private var force: Boolean = false
  /** How much logging output to print. Either none (default),
    * verbose or debug. */
  private var logging: Option[String] = None
  /** Which compilation phases should be logged during compilation. */
  private var logPhase: List[String] = Nil

  /** Whether to use implicit predefined values or not. */
  private var usepredefs: Option[Boolean] = None
  /** Instruct the compiler to generate debugging information */
  private var debugInfo: Option[String] = None
  /** Instruct the compiler to use additional parameters */
  private var addParams: String = ""
  /** Instruct the compiler to generate deprecation information. */
  private var deprecation: Option[Boolean] = None
  /** Instruct the compiler to generate unchecked information. */
  private var unchecked: Option[Boolean] = None

  /** Whether the compiler is being debuged. Prints more information in case
   *  in case of failure. */
  private var scalacDebugging: Boolean = false

/*============================================================================*\
**                             Properties setters                             **
\*============================================================================*/

  /** Sets the srcdir attribute. Used by Ant.
   *  @param input The value of <code>origin</code>. */
  def setSrcdir(input: Path) =
    if (origin.isEmpty) origin = Some(input)
    else origin.get.append(input)

  /** Sets the <code>origin</code> as a nested src Ant parameter.
   *  @return An origin path to be configured. */
  def createSrc(): Path = {
    if (origin.isEmpty) origin = Some(new Path(getProject()))
    origin.get.createPath()
  }

  /** Sets the <code>origin</code> as an external reference Ant parameter.
   *  @param input A reference to an origin path. */
  def setSrcref(input: Reference) =
    createSrc().setRefid(input)

  /** Sets the <code>destdir</code> attribute. Used by Ant.
   *  @param input The value of <code>destination</code>. */
  def setDestdir(input: File) =
      destination = Some(input)

  /** Sets the <code>classpath</code> attribute. Used by Ant.
   *  @param input The value of <code>classpath</code>. */
  def setClasspath(input: Path) =
    if (classpath.isEmpty) classpath = Some(input)
    else classpath.get.append(input)

  /** Sets the <code>classpath</code> as a nested classpath Ant parameter.
   *  @return A class path to be configured. */
  def createClasspath(): Path = {
    if (classpath.isEmpty) classpath = Some(new Path(getProject()))
    classpath.get.createPath()
  }

  /** Sets the <code>classpath</code> as an external reference Ant parameter.
   *  @param input A reference to a class path. */
  def setClasspathref(input: Reference) =
    createClasspath().setRefid(input)

  /** Sets the <code>sourcepath</code> attribute. Used by Ant.
   *  @param input The value of <code>sourcepath</code>. */
  def setSourcepath(input: Path) =
    if (sourcepath.isEmpty) sourcepath = Some(input)
    else sourcepath.get.append(input)

  /** Sets the <code>sourcepath</code> as a nested sourcepath Ant parameter.
   *  @return A source path to be configured. */
  def createSourcepath(): Path = {
    if (sourcepath.isEmpty) sourcepath = Some(new Path(getProject()))
    sourcepath.get.createPath()
  }

  /** Sets the <code>sourcepath</code> as an external reference Ant parameter.
   *  @param input A reference to a source path. */
  def setSourcepathref(input: Reference) =
    createSourcepath().setRefid(input)

  /** Sets the boot classpath attribute. Used by Ant.
   *
   *  @param input The value of <code>bootclasspath</code>. */
  def setBootclasspath(input: Path) =
    if (bootclasspath.isEmpty) bootclasspath = Some(input)
    else bootclasspath.get.append(input)

  /** Sets the <code>bootclasspath</code> as a nested sourcepath Ant
   *  parameter.
   *  @return A source path to be configured. */
  def createBootclasspath(): Path = {
    if (bootclasspath.isEmpty) bootclasspath = Some(new Path(getProject()))
    bootclasspath.get.createPath()
  }

  /** Sets the <code>bootclasspath</code> as an external reference Ant
   *  parameter.
   *  @param input A reference to a source path. */
  def setBootclasspathref(input: Reference) =
    createBootclasspath().setRefid(input)

  /** Sets the external extensions path attribute. Used by Ant.
   *  @param input The value of <code>extdirs</code>. */
  def setExtdirs(input: Path) =
    if (extdirs.isEmpty) extdirs = Some(input)
    else extdirs.get.append(input)

  /** Sets the <code>extdirs</code> as a nested sourcepath Ant parameter.
   *  @return An extensions path to be configured. */
  def createExtdirs(): Path = {
    if (extdirs.isEmpty) extdirs = Some(new Path(getProject()))
    extdirs.get.createPath()
  }

  /** Sets the <code>extdirs</code> as an external reference Ant parameter.
   *  @param input A reference to an extensions path. */
  def setExtdirsref(input: Reference) =
    createExtdirs().setRefid(input)

  /** Sets the <code>encoding</code> attribute. Used by Ant.
   *  @param input The value of <code>encoding</code>. */
  def setEncoding(input: String): Unit =
    encoding = Some(input)

  /** Sets the <code>target</code> attribute. Used by Ant.
   *  @param input The value for <code>target</code>. */
  def setTarget(input: String): Unit =
    if (Target.isPermissible(input)) backend = Some(input)
    else error("Unknown target '" + input + "'")

  /** Sets the <code>force</code> attribute. Used by Ant.
   *  @param input The value for <code>force</code>. */
  def setForce(input: Boolean): Unit =
    force = input

  /** Sets the logging level attribute. Used by Ant.
   *  @param input The value for <code>logging</code>. */
  def setLogging(input: String) =
    if (LoggingLevel.isPermissible(input)) logging = Some(input)
    else error("Logging level '" + input + "' does not exist.")

  /** Sets the <code>logphase</code> attribute. Used by Ant.
   *  @param input The value for <code>logPhase</code>. */
  def setLogPhase(input: String) = {
    logPhase = List.fromArray(input.split(",")).flatMap { s: String =>
      val st = s.trim()
      if (CompilerPhase.isPermissible(st))
        (if (input != "") List(st) else Nil)
      else {
        error("Phase " + st + " in log does not exist.")
        Nil
      }
    }
  }

  /** Sets the <code>usepredefs</code> attribute. Used by Ant.
   *  @param input The value for <code>usepredefs</code>. */
  def setUsepredefs(input: Boolean): Unit =
    usepredefs = Some(input)

  /** Set the <code>debug</code> info attribute.
   *  @param input The value for <code>debug</code>. */
  def setDebuginfo(input: String): Unit =
    debugInfo = Some(input)

  /** Set the <code>addparams</code> info attribute.
   *  @param input The value for <code>addparams</code>. */
  def setAddparams(input: String): Unit =
    addParams = input

  /** Set the <code>deprecation</code> info attribute.
   *  @param input One of the flags <code>yes/no</code> or <code>on/off</code>. */
  def setDeprecation(input: String): Unit =
    if (Flag.isPermissible(input))
      deprecation = Some("yes" == input || "on" == input)
    else
      error("Unknown deprecation flag '" + input + "'")

  /** Set the <code>unchecked</code> info attribute.
   *  @param input One of the flags <code>yes/no</code> or <code>on/off</code>. */
  def setUnchecked(input: String): Unit =
    if (Flag.isPermissible(input))
      unchecked = Some("yes" == input || "on" == input)
    else
      error("Unknown unchecked flag '" + input + "'")

  /** Set the <code>scalacdebugging</code> info attribute.
   *  @param input The specified flag */
  def setScalacdebugging(input: Boolean): Unit =
    scalacDebugging = input

/*============================================================================*\
**                             Properties getters                             **
\*============================================================================*/

  /** Gets the value of the <code>classpath</code> attribute in a
   *  Scala-friendly form.
   *  @return The class path as a list of files. */
  private def getClasspath: List[File] =
    if (classpath.isEmpty) error("Member 'classpath' is empty.")
    else List.fromArray(classpath.get.list()).map(nameToFile)

  /** Gets the value of the <code>origin</code> attribute in a
   *  Scala-friendly form.
   *  @return The origin path as a list of files. */
  private def getOrigin: List[File] =
    if (origin.isEmpty) error("Member 'origin' is empty.")
    else List.fromArray(origin.get.list()).map(nameToFile)

  /** Gets the value of the <code>destination</code> attribute in a
   *  Scala-friendly form.
   *  @return The destination as a file. */
  private def getDestination: File =
    if (destination.isEmpty) error("Member 'destination' is empty.")
    else existing(getProject().resolveFile(destination.get.toString()))

  /** Gets the value of the <code>sourcepath</code> attribute in a
   *  Scala-friendly form.
   *  @return The source path as a list of files. */
  private def getSourcepath: List[File] =
    if (sourcepath.isEmpty) error("Member 'sourcepath' is empty.")
    else List.fromArray(sourcepath.get.list()).map(nameToFile)

  /** Gets the value of the <code>bootclasspath</code> attribute in a
   *  Scala-friendly form.
   *  @return The boot class path as a list of files. */
  private def getBootclasspath: List[File] =
    if (bootclasspath.isEmpty) error("Member 'bootclasspath' is empty.")
    else List.fromArray(bootclasspath.get.list()).map(nameToFile)

  /** Gets the value of the <code>extdirs</code> attribute in a
   *  Scala-friendly form.
   *  @return The extensions path as a list of files. */
  private def getExtdirs: List[File] =
    if (extdirs.isEmpty) error("Member 'extdirs' is empty.")
    else List.fromArray(extdirs.get.list()).map(nameToFile)

/*============================================================================*\
**                       Compilation and support methods                      **
\*============================================================================*/

  /** This is forwarding method to circumvent bug #281 in Scala 2. Remove when
   *  bug has been corrected. */
  override protected def getDirectoryScanner(baseDir: File) =
    super.getDirectoryScanner(baseDir)

  /** Transforms a string name into a file relative to the provided base
   *  directory.
   *  @param base A file pointing to the location relative to which the name
   *              will be resolved.
   *  @param name A relative or absolute path to the file as a string.
   *  @return     A file created from the name and the base file. */
  private def nameToFile(base: File)(name: String): File =
    existing(fileUtils.resolveFile(base, name))

  /** Transforms a string name into a file relative to the build root
   *  directory.
   *  @param name A relative or absolute path to the file as a string.
   *  @return     A file created from the name. */
  private def nameToFile(name: String): File =
    existing(getProject().resolveFile(name))

  /** Tests if a file exists and prints a warning in case it doesn't. Always
   *  returns the file, even if it doesn't exist.
   *  @param file A file to test for existance.
   *  @return     The same file. */
  private def existing(file: File): File = {
    if (!file.exists())
      log("Element '" + file.toString() + "' does not exist.",
          Project.MSG_WARN)
    file
  }

  /** Transforms a path into a Scalac-readable string.
   *  @param path A path to convert.
   *  @return     A string-representation of the path like <code>a.jar:b.jar</code>. */
  private def asString(path: List[File]): String =
    path.map(asString).mkString(File.pathSeparator)

  /** Transforms a file into a Scalac-readable string.
   *  @param path A file to convert.
   *  @return     A string-representation of the file like <code>/x/k/a.scala</code>. */
  private def asString(file: File): String =
    file.getAbsolutePath()

  /** Generates a build error. Error location will be the current task in the
   *  ant file.
   *  @param message         A message describing the error.
   *  @throws BuildException A build error exception thrown in every case. */
  private def error(message: String): Nothing =
    throw new BuildException(message, getLocation())

/*============================================================================*\
**                           The big execute method                           **
\*============================================================================*/

  /** Initializes settings and source files */
  protected def initialize: Pair[Settings, List[File]] = {
    // Tests if all mandatory attributes are set and valid.
    if (origin.isEmpty) error("Attribute 'srcdir' is not set.")
    if (getOrigin.isEmpty) error("Attribute 'srcdir' is not set.")
    if (!destination.isEmpty && !destination.get.isDirectory())
      error("Attribute 'destdir' does not refer to an existing directory.")
    if (destination.isEmpty) destination = Some(getOrigin.head)

    val mapper = new GlobPatternMapper()
    mapper.setTo("*.class")
    mapper.setFrom("*.scala")

    // Scans source directories to build up a compile lists.
    // If force is false, only files were the .class file in destination is
    // older than the .scala file will be used.
    val sourceFiles: List[File] =
      for {
        val originDir <- getOrigin
        val originFile <- {
          var includedFiles =
            getDirectoryScanner(originDir).getIncludedFiles()
          if (!force) {
            includedFiles = new SourceFileScanner(this).
              restrict(includedFiles, originDir, destination.get, mapper)
          }
          val list = List.fromArray(includedFiles)
          if (scalacDebugging && list.length > 0)
            log(
              list.mkString(
                "Compiling source file" +
                (if (list.length > 1) "s: " else ": "),
                ", ",
                " "
              ) + "to " + getDestination.toString()
            )
          else if (list.length > 0)
            log(
              "Compiling " + list.length + " source file" +
              (if (list.length > 1) "s" else "") +
              (" to " + getDestination.toString())
            )
          else
            log("No files selected for compilation", Project.MSG_VERBOSE)

          list
        }
      }
      yield {
        log(originFile.toString(), Project.MSG_DEBUG)
        nameToFile(originDir)(originFile)
      }

    // Builds-up the compilation settings for Scalac with the existing Ant
    // parameters.
    val settings = new Settings(error)
    settings.outdir.value = asString(destination.get)
    if (!classpath.isEmpty)
      settings.classpath.value = asString(getClasspath)
    if (!sourcepath.isEmpty)
      settings.sourcepath.value = asString(getSourcepath)
    else if (origin.get.size() > 0)
      settings.sourcepath.value = origin.get.list()(0)
    if (!bootclasspath.isEmpty)
      settings.bootclasspath.value = asString(getBootclasspath)
    if (!extdirs.isEmpty) settings.extdirs.value = asString(getExtdirs)
    if (!encoding.isEmpty) settings.encoding.value = encoding.get
    if (!backend.isEmpty) settings.target.value = backend.get
    if (!logging.isEmpty && logging.get == "verbose")
      settings.verbose.value = true
    else if (!logging.isEmpty && logging.get == "debug") {
      settings.verbose.value = true
      settings.debug.value = true
    }
    if (!logPhase.isEmpty) settings.log.value = logPhase
    if (!usepredefs.isEmpty) settings.nopredefs.value = !usepredefs.get
    if (!debugInfo.isEmpty) settings.debuginfo.value = debugInfo.get
    if (!deprecation.isEmpty) settings.deprecation.value = deprecation.get
    if (!unchecked.isEmpty) settings.unchecked.value = unchecked.get

    log("Scalac params = '" + addParams + "'", Project.MSG_DEBUG)
    var args =
      if (addParams.trim() == "") Nil
      else List.fromArray(addParams.trim().split(" ")).map(.trim())
    while (!args.isEmpty) {
      val argsBuf = args
      if (args.head startsWith "-") {
        for (setting <- settings.allSettings)
          args = setting.tryToSet(args);
      }
      else error("Parameter '" + args.head + "' does not start with '-'.")
      if (argsBuf eq args)
        error("Parameter '" + args.head + "' is not recognised by Scalac.")
    }
    Pair(settings, sourceFiles)
  }

  /** Performs the compilation. */
  override def execute() = {
    val Pair(settings, sourceFiles) = initialize
    val reporter = new ConsoleReporter(settings)

    // Compiles the actual code
    val compiler = new Global(settings, reporter)
    try {
      (new compiler.Run).compile(sourceFiles.map (.toString()))
     }
     catch {
      case exception: Throwable if (exception.getMessage ne null) =>
        exception.printStackTrace()
        error("Compile failed because of an internal compiler error (" +
          exception.getMessage + "); see the error output for details.")
      case exception =>
        exception.printStackTrace()
        error("Compile failed because of an internal compiler error " +
              "(no error message provided); see the error output for details.")
    }
    reporter.printSummary()
    if (reporter.hasErrors)
      error(
          "Compile failed with " +
          reporter.ERROR.count + " error" +
          (if (reporter.ERROR.count > 1) "s" else "") +
          "; see the compiler error output for details.")
    else if (reporter.WARNING.count > 0)
      log(
          "Compile suceeded with " +
          reporter.WARNING.count + " warning" +
          (if (reporter.WARNING.count > 1) "s" else "") +
          "; see the compiler output for details.")
  }

}
